src/libraries/Microsoft.PowerFx.Core/Public/Values/UntypedObjectValue.cs

```diff
@@ -79,5 +79,33 @@ public override void ToExpression(StringBuilder sb, FormulaValueSerializerSettin
             // Not supported for the time being.
             throw new NotImplementedException("UntypedObjectValue cannot be serialized.");
         }
+    }
+
+    public abstract class UntypedObjectBase : IUntypedObject
+    {
+        public abstract IUntypedObject this[int index] { get; }
+
+        public abstract FormulaType Type { get; }
+
+        public abstract int GetArrayLength();
+
+        public abstract bool GetBoolean();
+
+        public abstract decimal GetDecimal();
+
+        public abstract double GetDouble();
+
+        public abstract string GetString();
+
+        public abstract string GetUntypedNumber();
+
+        public abstract bool TryGetProperty(string value, out IUntypedObject result);
+
+        public abstract bool TryGetPropertyNames(out IEnumerable<string> propertyNames);
+
+        public virtual bool TrySetProperty(string name, FormulaValue value)
+        {
+            throw new NotImplementedException();
+        }

```

Comment:
Let's just put this on UntypedObjectValue 

 and avoid the new class.

Reply:
after further exploration - that's too much churn. 
This approach is probably our best bet.